File: docs/workstreams/lithic-pq-security/remediation/reference/authorization_state_transition_v1.py
"""Independent Python AUTHORIZATION_STATE_TRANSITION_V1 conformance runner."""
from __future__ import annotations

import copy
import json
from pathlib import Path

VECTORS = Path(__file__).resolve().parents[1] / "vectors" / "authorization_state_transitions.json"
MIN_CANCELLATION_WINDOW = 10
MIN_ACTIVATION_DELAY = 11
MAX_PENDING_HORIZON = 100_000


class Reject(Exception):
    pass


def key(state: dict, slot: int, epoch: int) -> dict | None:
    return next((item for item in state["keys"] if item["slot"] == slot and item["epoch"] == epoch), None)


def next_epoch(state: dict, slot: int) -> int:
    epochs = [item["epoch"] for item in state["keys"] if item["slot"] == slot]
    return max(epochs, default=0) + 1


def request(state: dict, op: dict, consensus_height: int) -> None:
    if state["pending"] is not None:
        raise Reject("PENDING_EXISTS")
    kind = op["kind"]
    if kind not in {"REGISTER", "ROTATE", "DISABLE", "RECOVER"}:
        raise Reject("UNKNOWN_OPERATION")
    if op["deadline"] < consensus_height + MIN_CANCELLATION_WINDOW:
        raise Reject("CANCELLATION_WINDOW_TOO_SHORT")
    if op["activation"] <= op["deadline"] or op["activation"] < consensus_height + MIN_ACTIVATION_DELAY:
        raise Reject("INVALID_WINDOW")
    if op["activation"] > consensus_height + MAX_PENDING_HORIZON:
        raise Reject("PENDING_HORIZON_EXCEEDED")

    prior_epoch = op.get("prior_epoch", 0)
    if kind in {"REGISTER", "ROTATE", "RECOVER"} and op["epoch"] != next_epoch(state, op["slot"]):
        raise Reject("NON_MONOTONIC_EPOCH")
    if kind == "REGISTER" and prior_epoch != 0:
        raise Reject("INVALID_PREDECESSOR")
    if kind in {"ROTATE", "DISABLE"}:
        prior = key(state, op["slot"], prior_epoch)
        if prior is None or prior["state"] != "ACTIVE":
            raise Reject("NO_ACTIVE_PREDECESSOR")
    if kind == "RECOVER":
        prior = key(state, op["prior_slot"], prior_epoch)
        if prior is None or prior["state"] != "ACTIVE":
            raise Reject("NO_ACTIVE_PREDECESSOR")

    state["pending"] = {
        "kind": kind,
        "slot": op["slot"],
        "epoch": op["epoch"],
        "prior_epoch": prior_epoch,
        "prior_slot": op.get("prior_slot", 0 if kind == "REGISTER" else op["slot"]),
        "deadline": op["deadline"],
        "activation": op["activation"],
        "next_policy_version": state["policy_version"] + 1,
        "next_authorization_epoch": state["authorization_epoch"] + 1,
    }
    if kind != "DISABLE":
        state["keys"].append({"slot": op["slot"], "epoch": op["epoch"], "state": "PENDING_ACTIVATION"})
    state["records"].append(f"{kind}_REQUESTED")


def cancel(state: dict, op: dict, consensus_height: int) -> None:
    pending = state["pending"]
    if pending is None:
        raise Reject("NO_PENDING_MUTATION")
    recovery_cancel = op.get("authority", "ordinary") == "recovery" and consensus_height < pending["activation"]
    if consensus_height > pending["deadline"] and not recovery_cancel:
        raise Reject("CANCELLATION_CLOSED")
    if pending["kind"] != "DISABLE":
        proposed = key(state, pending["slot"], pending["epoch"])
        if proposed is None or proposed["state"] != "PENDING_ACTIVATION":
            raise Reject("PENDING_KEY_MISMATCH")
        proposed["state"] = "CANCELLED"
    state["pending"] = None
    state["records"].append("PENDING_CANCELLED")


def activate(state: dict, op: dict, consensus_height: int) -> None:
    pending = state["pending"]
    if pending is None:
        raise Reject("NO_PENDING_MUTATION")
    if "claimed_execution_height" in op:
        raise Reject("UNREGISTERED_EXECUTION_HEIGHT_FIELD")
    if consensus_height < pending["activation"] or consensus_height <= pending["deadline"]:
        raise Reject("NOT_MATURE")
    kind = pending["kind"]
    if kind == "REGISTER":
        key(state, pending["slot"], pending["epoch"])["state"] = "ACTIVE"
    elif kind == "ROTATE":
        key(state, pending["slot"], pending["prior_epoch"])["state"] = "ROTATED"
        key(state, pending["slot"], pending["epoch"])["state"] = "ACTIVE"
    elif kind == "DISABLE":
        key(state, pending["slot"], pending["prior_epoch"])["state"] = "DISABLED"
    elif kind == "RECOVER":
        key(state, pending["prior_slot"], pending["prior_epoch"])["state"] = "RECOVERED"
        key(state, pending["slot"], pending["epoch"])["state"] = "ACTIVE"
    else:
        raise Reject("UNKNOWN_OPERATION")
    state["policy_version"] = pending["next_policy_version"]
    state["authorization_epoch"] = pending["next_authorization_epoch"]
    state["pending"] = None
    state["records"].append("PENDING_ACTIVATED")


def governance(state: dict, op: dict, consensus_height: int) -> None:
    if op["action"] != op["inner_action"] or op["action"] != 4 or op["target_type"] != 2 or op["emergency"] is not True:
        raise Reject("GOVERNANCE_DISPATCH_MISMATCH")
    if op["authority"] != "PQ_REGISTRY_EMERGENCY_2_OF_3":
        raise Reject("WRONG_AUTHORITY_ROOT")
    state["records"].append("EMERGENCY_DISABLE_AUTHORIZED")


def run_case(case: dict) -> dict:
    state = copy.deepcopy(case["initial"])
    state.setdefault("pending", None)
    state.setdefault("records", [])
    accepted, errors = [], []
    handlers = {"request": request, "cancel": cancel, "activate": activate, "governance": governance}
    for step in case["operations"]:
        before = copy.deepcopy(state)
        operation = {key: value for key, value in step.items() if key != "height"}
        consensus_height = step["height"]
        try:
            handlers[operation["type"]](state, operation, consensus_height)
        except Reject as exc:
            state = before
            accepted.append(False)
            errors.append(str(exc))
        else:
            accepted.append(True)
            errors.append("OK")
    state["keys"] = sorted(state["keys"], key=lambda item: (item["slot"], item["epoch"]))
    return {"accepted": accepted, "errors": errors, **state}


def main() -> None:
    vectors = json.loads(VECTORS.read_text(encoding="utf-8"))
    if vectors.get("schema") != "AUTHORIZATION_STATE_TRANSITION_V1":
        raise SystemExit("wrong vector schema")
    for case in vectors["cases"]:
        actual = run_case(case)
        if actual != case["expected"]:
            raise SystemExit(f"state transition mismatch: {case['name']}\nexpected={case['expected']}\nactual={actual}")
    print(f"python authorization state machine verified {len(vectors['cases'])} complete sequences")


if __name__ == "__main__":
    main()

File: docs/workstreams/lithic-pq-security/remediation/reference/authorization_state_transition_v1.mjs
// Independent JavaScript AUTHORIZATION_STATE_TRANSITION_V1 conformance runner.
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const file = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..', 'vectors', 'authorization_state_transitions.json');
const vectors = JSON.parse(fs.readFileSync(file, 'utf8'));
const MIN_CANCELLATION_WINDOW = 10;
const MIN_ACTIVATION_DELAY = 11;
const MAX_PENDING_HORIZON = 100000;
class Rejected extends Error {}
const clone = value => JSON.parse(JSON.stringify(value));
const findKey = (state, slot, epoch) => state.keys.find(item => item.slot === slot && item.epoch === epoch);
const nextEpoch = (state, slot) => Math.max(0, ...state.keys.filter(item => item.slot === slot).map(item => item.epoch)) + 1;

function request(state, op, consensusHeight) {
  if (state.pending !== null) throw new Rejected('PENDING_EXISTS');
  if (!['REGISTER', 'ROTATE', 'DISABLE', 'RECOVER'].includes(op.kind)) throw new Rejected('UNKNOWN_OPERATION');
  if (op.deadline < consensusHeight + MIN_CANCELLATION_WINDOW) throw new Rejected('CANCELLATION_WINDOW_TOO_SHORT');
  if (op.activation <= op.deadline || op.activation < consensusHeight + MIN_ACTIVATION_DELAY) throw new Rejected('INVALID_WINDOW');
  if (op.activation > consensusHeight + MAX_PENDING_HORIZON) throw new Rejected('PENDING_HORIZON_EXCEEDED');
  const prior = op.prior_epoch ?? 0;
  if (['REGISTER', 'ROTATE', 'RECOVER'].includes(op.kind) && op.epoch !== nextEpoch(state, op.slot)) throw new Rejected('NON_MONOTONIC_EPOCH');
  if (op.kind === 'REGISTER' && prior !== 0) throw new Rejected('INVALID_PREDECESSOR');
  if (['ROTATE', 'DISABLE'].includes(op.kind)) {
    const old = findKey(state, op.slot, prior);
    if (!old || old.state !== 'ACTIVE') throw new Rejected('NO_ACTIVE_PREDECESSOR');
  }
  if (op.kind === 'RECOVER') {
    const old = findKey(state, op.prior_slot, prior);
    if (!old || old.state !== 'ACTIVE') throw new Rejected('NO_ACTIVE_PREDECESSOR');
  }
  state.pending = {kind: op.kind, slot: op.slot, epoch: op.epoch, prior_epoch: prior, prior_slot: op.prior_slot ?? (op.kind === 'REGISTER' ? 0 : op.slot), deadline: op.deadline, activation: op.activation, next_policy_version: state.policy_version + 1, next_authorization_epoch: state.authorization_epoch + 1};
  if (op.kind !== 'DISABLE') state.keys.push({slot: op.slot, epoch: op.epoch, state: 'PENDING_ACTIVATION'});
  state.records.push(`${op.kind}_REQUESTED`);
}

function cancel(state, op, consensusHeight) {
  const pending = state.pending;
  if (!pending) throw new Rejected('NO_PENDING_MUTATION');
  const recoveryCancel = (op.authority ?? 'ordinary') === 'recovery' && consensusHeight < pending.activation;
  if (consensusHeight > pending.deadline && !recoveryCancel) throw new Rejected('CANCELLATION_CLOSED');
  if (pending.kind !== 'DISABLE') {
    const proposed = findKey(state, pending.slot, pending.epoch);
    if (!proposed || proposed.state !== 'PENDING_ACTIVATION') throw new Rejected('PENDING_KEY_MISMATCH');
    proposed.state = 'CANCELLED';
  }
  state.pending = null;
  state.records.push('PENDING_CANCELLED');
}

function activate(state, op, consensusHeight) {
  const pending = state.pending;
  if (!pending) throw new Rejected('NO_PENDING_MUTATION');
  if ('claimed_execution_height' in op) throw new Rejected('UNREGISTERED_EXECUTION_HEIGHT_FIELD');
  if (consensusHeight < pending.activation || consensusHeight <= pending.deadline) throw new Rejected('NOT_MATURE');
  if (pending.kind === 'REGISTER') findKey(state, pending.slot, pending.epoch).state = 'ACTIVE';
  else if (pending.kind === 'ROTATE') {
    findKey(state, pending.slot, pending.prior_epoch).state = 'ROTATED';
    findKey(state, pending.slot, pending.epoch).state = 'ACTIVE';
  } else if (pending.kind === 'DISABLE') findKey(state, pending.slot, pending.prior_epoch).state = 'DISABLED';
  else if (pending.kind === 'RECOVER') {
    findKey(state, pending.prior_slot, pending.prior_epoch).state = 'RECOVERED';
    findKey(state, pending.slot, pending.epoch).state = 'ACTIVE';
  } else throw new Rejected('UNKNOWN_OPERATION');
  state.policy_version = pending.next_policy_version;
  state.authorization_epoch = pending.next_authorization_epoch;
  state.pending = null;
  state.records.push('PENDING_ACTIVATED');
}

function governance(state, op, consensusHeight) {
  if (op.action !== op.inner_action || op.action !== 4 || op.target_type !== 2 || op.emergency !== true) throw new Rejected('GOVERNANCE_DISPATCH_MISMATCH');
  if (op.authority !== 'PQ_REGISTRY_EMERGENCY_2_OF_3') throw new Rejected('WRONG_AUTHORITY_ROOT');
  state.records.push('EMERGENCY_DISABLE_AUTHORIZED');
}

const handlers = {request, cancel, activate, governance};
function run(test) {
  let state = clone(test.initial);
  state.pending ??= null;
  state.records ??= [];
  const accepted = [], errors = [];
  for (const step of test.operations) {
    const before = clone(state);
    const consensusHeight = step.height;
    const operation = Object.fromEntries(Object.entries(step).filter(([key]) => key !== 'height'));
    try {
      handlers[operation.type](state, operation, consensusHeight);
      accepted.push(true); errors.push('OK');
    } catch (error) {
      if (!(error instanceof Rejected)) throw error;
      state = before; accepted.push(false); errors.push(error.message);
    }
  }
  state.keys.sort((a, b) => a.slot - b.slot || a.epoch - b.epoch);
  return {accepted, errors, ...state};
}

if (vectors.schema !== 'AUTHORIZATION_STATE_TRANSITION_V1') throw new Error('wrong vector schema');
for (const test of vectors.cases) {
  const actual = run(test);
  if (JSON.stringify(actual) !== JSON.stringify(test.expected)) throw new Error(`state transition mismatch: ${test.name}\nexpected=${JSON.stringify(test.expected)}\nactual=${JSON.stringify(actual)}`);
}
process.stdout.write(`javascript authorization state machine verified ${vectors.cases.length} complete sequences\n`);
